ggml-cuda.cu

```diff
@@ -1960,20 +1960,84 @@ static void ggml_cuda_mul_mat(ggml_backend_cuda_context & ctx, const ggml_tensor
     }
 }
 
+struct mmid_row_mapping {
+    int64_t i1;
+    int64_t i2;
+};
+
+static __global__ void k_copy_src1_to_contiguous(const char * src1_original, char * src1_contiguous,
+                                                 int * cur_src1_row, mmid_row_mapping * row_mapping,
+                                                 const char * ids_dev, int64_t i02, int64_t ids_nb1, int64_t ids_nb0,
```

Comment:
```suggestion
static __global__ void k_copy_src1_to_contiguous(const char * __restrict__ src1_original, char * __restrict__ src1_contiguous,
                                                 int * __restrict__ cur_src1_row, mmid_row_mapping * __restrict__ row_mapping,
                                                 const char * __restrict__ ids_dev, int64_t i02, int64_t ids_nb1, int64_t ids_nb0,
```

Especially on older hardware adding `__restrict__` when the pointers are not aliased can make the code much faster.

Also, avoid the use of `int64_t` unless you really need to; `int` is faster since CUDA uses 32 bit registers.